### DISCUSSION TITLE
Using calibrated cameras off of Meta Quest 3 (and general development from MSFT)

### DISCUSSION THREAD

**Discussion by llfab** [StereoKit/StereoKit#872] [Q&A]
Sorry, that this is not 100% related to StereoKit: we are creating medical devices and typically want to do our own optical pattern tracking using the calibrated cameras of the goggle. We had done that for HL2 using it in Dev Mode. 

As we believe HL2 is kind of deprecated we will be playing with Apple Vision Pro and Meta Quest 3. From what we find, there seems to be no way to access the camera stream(s) of the inside out cameras with calibration info to throw that into our pattern tracking algos. 

So, some questions hoping you are much closer connected to things (@maluoi):
- Would you currently consider MSFT out of the picture regarding future MR goggles?
- Do you see any kind of (dirty) workaround for getting inside out camera stream(s) with their calibration info for either Apple Vision Pro or Meta Quest 3 or any other relevant competitor?

Thanks,
Fabian

**Comment by maluoi** [comment]
## Microsoft's future

UWP was dead shortly after HL2 shipped, and is officially deprecated at this point. The WMR runtime for Desktop has an official support End of Life in... 2026? But is being removed from Windows in June/July of this year. I wouldn't really consider Microsoft to be 100% out, but maybe 90% out. If we see another consumer/business headset from Microsoft (and I think there's still a decent chance we may), my guesses are that it will be Android, and that Microsoft will not be directly making it.

I'm honestly curious how Microsoft is going to pull off the military AR headset contract at this point, given how much talent they've bled. But it's a big commitment that may force them to stay in the game just a little bit.

## Camera Streams

Apple Vision Pro, I expect Apple would chase you with lawyers if you found a way to access their cameras. Meta, given their mass consumer focus, I don't think they're likely to expose camera streams unless it becomes an industry standard, or more socially acceptable.

The story gets a bit better with some other headsets though. Snapdragon Spaces headsets have an `XR_QCOMX_camera_frame_access` OpenXR extension, [which you can find some additional details about over here (Spaces SDK for Unity docs)](https://docs.spaces.qualcomm.com/unity/samples/CameraFrameAccessSample.html). This would be devices like Lynx R1, Lenovo ThinkReality A3, or Lenovo VRX. I don't have any direct experience with this extension, and it does look like this may be more webcam than tracking cam frames? But it does have extrinsics/intrinsics.

Tracking camera data is something that may become less and less available over time though? I believe there's a lot of advantages to having dedicated hardware for processing tracking data, passthrough, stuff that's really expensive yet critical to run uninterrupted. Stuff like the HL2's Holographic Processing Unit, or AVP's R1 chip. It's a _lot_ of data to be shuffling back and forth to the main CPUs, if you have to do that, which is why HL2 had a major performance penalty for research mode.

**Reply by llfab** [reply]
Puhh, thank you so much Nick for this extensive explanation. Too bad, but the arguments sound sound...

(love your work!)

**Reply by sophia-anais** [reply]
Hi Nick, 

we really appreciate your detailed answer, this is very helpful!
Just another question: I guess our use case is somewhat special as we want to track objects continously e.g. based on markers like QR codes attached to them, even with those objects being in motion. 
In your experience, is there any solution out there to track positions without directly accessing the raw camera feed?

Thanks a lot in advance!
Sophia

**Reply by maluoi** [reply]
For marker tracking, there's a couple OpenXR extensions depending on your headset! The industry hasn't standardized this yet, but given how many there, I certainly hope they will soon 😅

[XR_VARJO_marker_tracking](https://registry.khronos.org/OpenXR/specs/1.0/html/xrspec.html#XR_VARJO_marker_tracking) - Aruco markers
[XR_ML_marker_understanding](https://registry.khronos.org/OpenXR/specs/1.0/html/xrspec.html#XR_ML_marker_understanding) - Aruco + QR codes + AprilTag
[XR_MSFT_scene_marker](https://registry.khronos.org/OpenXR/specs/1.0/html/xrspec.html#XR_MSFT_scene_marker) - QR codes

These two aren't in the public OpenXR spec yet, but have samples implemented in Unity and Unreal.
[XR_QCOMX_marker_tracking](https://docs.spaces.qualcomm.com/unity/samples/QRCodeTrackingSample.html) - QR codes
[XR_QCOM_image_tracking](https://docs.spaces.qualcomm.com/unity/samples/ImageTrackingSample.html) - Images

I'm only really directly familiar with the MSFT one myself, and it did have some caveats to it! Microsoft's QR code tracking was _really_ accurate, but only would update once per second, and required you to be either quite close to the QR code, or have a really big QR code! It was using the tracking cameras for this, which were fairly low resolution.

Other systems may come with some other caveats. Like in general I'd expect Aruco markers or AprilTags to be kinda faster than others, or some systems might require markers to be fixed size, or might need to know about the image/id in advance.

**Reply by sophia-anais** [reply]
Thanks a lot for your answer! 😊
